src/algebra/pointwise.lean

```diff
@@ -157,151 +140,180 @@ begin
+protected def set_semiring.down (s : set_semiring α) : set α := s
+@[simp] protected lemma down_up {s : set α} : s.up.down = s := rfl
+@[simp] protected lemma up_down {s : set_semiring α} : s.down.up = s := rfl
+
+instance set_semiring.semiring [monoid α] : semiring (set_semiring α) :=
+{ add := λ s t, (s ∪ t : set α),
+  zero := (∅ : set α),
+  add_assoc := union_assoc,
+  zero_add := empty_union,
+  add_zero := union_empty,
+  add_comm := union_comm,
+  zero_mul := λ s, empty_mul,
+  mul_zero := λ s, mul_empty,
+  left_distrib := λ _ _ _, mul_union,
+  right_distrib := λ _ _ _, union_mul,
+  ..set.monoid }
+
+instance set_semiring.comm_semiring [comm_monoid α] : comm_semiring (set_semiring α) :=
+{ ..set.comm_monoid, ..set_semiring.semiring }
 
 /-- A multiplicative action of a monoid on a type β gives also a
  multiplicative action on the subsets of β. -/
 def smul_set_action [monoid α] [mul_action α β] :
   mul_action α (set β) :=
-{ smul     := λ a s, a • s,
-  mul_smul := λ a b s, set.ext $ λ x, iff.intro
-    (λ ⟨_, hy, _⟩, ⟨b • _, smul_mem_smul_set _ hy, by rwa ←mul_smul⟩)
-    (λ ⟨_, hy, _⟩, let ⟨_, hz, h⟩ := (mem_smul_set _ _ _).2 hy in
-      ⟨_, hz, by rwa [mul_smul, ←h]⟩),
+{ mul_smul := λ a b s, set.ext $ λ x, iff.intro
+    (λ ⟨_, hy, _⟩, ⟨b • _, smul_mem_smul_set hy, by rwa ←mul_smul ⟩)
+    (λ ⟨_, ⟨_, hz, h⟩, h2⟩, ⟨_, hz, by rwa [mul_smul, h, h2] ⟩),
   one_smul := λ b, set.ext $ λ x, iff.intro
-    (λ ⟨_, _, h⟩, by { rw [one_smul] at h; rwa h })
-    (λ h, ⟨_, h, by rw one_smul⟩) }
+    (λ ⟨_, _, h⟩, by { rw [one_smul] at h, rwa ← h })
+    (λ h, ⟨_, h, by rw one_smul ⟩),
+  ..set.has_scalar_set }
 
 section is_mul_hom
 open is_mul_hom
 
 variables [has_mul α] [has_mul β] (m : α → β) [is_mul_hom m]
 
 @[to_additive]
-lemma image_pointwise_mul (s t : set α) : m '' (s * t) = m '' s * m '' t :=
+lemma image_mul : m '' (s * t) = m '' s * m '' t :=
 set.ext $ assume y,
 begin
   split,
   { rintros ⟨_, ⟨_, _, _, _, rfl⟩, rfl⟩,
-    refine ⟨_, mem_image_of_mem _ ‹_›, _, mem_image_of_mem _ ‹_›, map_mul _ _ _⟩ },
-  { rintros ⟨_, ⟨_, _, rfl⟩, _, ⟨_, _, rfl⟩, rfl⟩,
-    refine ⟨_, ⟨_, ‹_›, _, ‹_›, rfl⟩, map_mul _ _ _⟩ }
+    exact ⟨_, _, mem_image_of_mem _ ‹_›, mem_image_of_mem _ ‹_›, (map_mul _ _ _).symm ⟩ },
+  { rintros ⟨_, _, ⟨_, _, rfl⟩, ⟨_, _, rfl⟩, rfl⟩, exact ⟨_, ⟨_, _, ‹_›, ‹_›, rfl⟩, map_mul _ _ _⟩ }
 end
 
 @[to_additive]
-lemma preimage_pointwise_mul_preimage_subset (s t : set β) : m ⁻¹' s * m ⁻¹' t ⊆ m ⁻¹' (s * t) :=
-begin
-  rintros _ ⟨_, _, _, _, rfl⟩,
-  exact ⟨_, ‹_›, _, ‹_›, map_mul _ _ _⟩,
-end
+lemma preimage_mul_preimage_subset {s t : set β} : m ⁻¹' s * m ⁻¹' t ⊆ m ⁻¹' (s * t) :=
+by { rintros _ ⟨_, _, _, _, rfl⟩, exact ⟨_, _, ‹_›, ‹_›, (map_mul _ _ _).symm ⟩ }
 
 end is_mul_hom
 
-variables [monoid α] [monoid β] [is_monoid_hom f]
+variables [monoid α] [monoid β]
 
-lemma pointwise_mul_image_is_semiring_hom : is_semiring_hom (image f) :=
+lemma pointwise_mul_image_is_semiring_hom (f : α → β) [is_monoid_hom f] :
```

Comment:
Do we want to use `is_monoid_hom` or `→*`?

Reply:
Good idea, I'll also change this.